### PR TITLE
Add authentication and access control

### PR DESCRIPTION
**Pull request by webb-ben**
This PR is a first pass on #20, adding simple access control to wis2box. This includes the an auth flask app which will become [wis2box-auth](https://github.com/wmo-im/wis2box-auth), wis2box CLI utility to configure access control, documentation, and docker-compose/nginx configured to use the auth server.

In its current configuration, the nginx makes a subrequest to the auth server for any request to the WAF or OAPI. The topic hierarchy is extracted out of the original request URI. If the topic is open, access will always be granted. If access control has been configured, the token will be authenticated.

Tokens are only required for topics with access control configured. In this iteration, tokens can be used to authenticate by appending them as arguments to the URI (i.e. `http://localhost:8999/pygeoapi/collections/data.core.observations-surface-land.mw.FWCL.landFixed/?token=[TOKEN]`). It will be important to make sure the token can also be passed along as a bearer token. 

File-by-file:
- **`.github/workflows/tests-docker.yml`:** At the moment, wis2box-auth uses the same parent image as wis2box. This is only a problem for the GitHub actions workflow which does not have tty. These changes more closely mirror the behavior of wis2box-ctl.py
- **`docker/auth/Dockerfile` & `docker/auth/app.py` & `docker/auth/entrypoint.sh`:** wis2box-auth repo. Needs to have access to a python version of wis2box that has auth, flask and gunicorn. Current solution is to have the Dockerfile piggyback on the wis2box image. Could use some cleanup.
- **`docker/default.env` & `wis2box/env.py`:** wis2box-auth store location. 
- **`docker/docker-compose.yml` & `docker/docker-compose.override.yml`:** Add wis2box-auth to docker compose deployment.
- **`docker/nginx/nginx.conf`:** Use wis2box-auth for authentication based on subrequest result.
- **`docs/source/index.rst` & `docs/source/auth.rst`:** Add documentation for auth
- **`wis2box/__init__.py`:** Add auth commands to wis2box CLI
- **`wis2box/auth/__init__.py`:** wis2box authentication CLI and core methods (`is_resource_open`, `is_token_authorized`) utilized by wis2box-auth.
- **`wis2box/auth/base.py`:** wis2box base auth provider using a sqlite as the datastore for passwords.

### CODE REVIEW

**Comment by petersilva** [comment]
question: Why isn't the separator in "data.core.observations-surface-land.mw.FWCL.landFixed" a / ? if people want to use MQTT (which uses / as a separator), then it will be far more consistent to use the same topic separator in all realms, and it will also look much more fashionable (REST I guess rest is old news by now) if / is used rather than period.

**Comment by tomkralidis** [comment]
@petersilva in which context? I agree we should keep with / as the separator, but it will not work in some spots. The API is one of them (the OGC APIs currently have a fixed structure / pathing).

**Comment by petersilva** [comment]
sounds like the OGC-API needs to understand what a base URL is.

**Comment by petersilva** [comment]
It's not important it was just a question... for my own understanding.

**Comment by petersilva** [comment]
question... on noaa sites, the keyword used is "bearer_token" rather than just "token" ...  Canadian stack has been built to support the former... we can add the latter... I'm just wondering what's "normal".

**Comment by petersilva** [comment]
so... I don't know what sort of review we want... 
* If it's looking over the code... it looks well done and and generally excellent.  I'll bless it immediately.
* if it's ensuring it works, requiring me to actually run it and look for bobos... that will take a couple of days for me to get to, before I can do that...
* If it's agreeing with the approach, then I look at how topics are being extracted:
```

def extract_topic(url: str) -> str:
￼    pattern = r'(?!^.*\/)(([a-z]{4})([\/.]{1})([a-z]{4})([\/.]{1})([a-z-]+)([\/.]{1})([a-z]{2})([\/.]{1})([A-Z]{4})([\/.]{1})([a-zA-Z]*))' # noqa
￼    prog = re.compile(pattern)
￼    match = prog.search(url)
￼    return match[0] if match else ''
```

extract_topic being based on a regex applied to the URL seems strange and fragile to me.  I really would prefer a baseURL approach that would remove the beginning, and then os.dirname() to remove the end.  That would be robust to changes of topic delimiters as well as conceptually simpler... but I don't think it's my call... 

please let me know which perspective I should use for this review.

**Comment by tomkralidis** [comment]
@petersilva minimum requirement is a code review.  If you have the time to check out the branch and test locally, that would be even better.  For now, code review works/is fine.

I agree on the regex for the auth and have discussed @webb-ben.  Given the WMO topic hierarchies are in development, the ideal way forward would be to have a dedicated package for WMO topic validation, which would be used various containers in wis2box, and any other workflow for that matter.  We are in parallel with WIS 2.0 specification development, and it's expected to have some ebb and flow / iterations between the two.

tl; dr; @webb-ben add a `TODO` in the auth container to move away from the regex matching over time.

**Review by petersilva — APPROVED** [review]
code looks good.

**Comment by david-i-berry** [comment]
I've just looked at the regex for topic extraction above

````python
def extract_topic(url: str) -> str:
    pattern = r'(?!^.*\/)(([a-z]{4})([\/.]{1})([a-z]{4})([\/.]{1})([a-z-]+)([\/.]{1})([a-z]{2})([\/.]{1})([A-Z]{4})([\/.]{1})([a-zA-Z]*))' # noqa
    prog = re.compile(pattern)
    match = prog.search(url)
    return match[0] if match else ''
````

The  second group looks like it should match ``core`` (or any other 4 letter word) as in ``data.core.xxxxx``, what about recommended data?

File: docker/auth/app.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+def authorize():
+    request_uri = request.headers.get('X-Original-URI')
+    request_ = request.from_values(request_uri)
+
+    LOGGER.debug('Extracting topic from request URI')
+    resource = extract_topic(request_uri)
+
+    LOGGER.debug('Extracting API token')
+    api_key = request_.args.get('token')
+    auth_header = request.headers.get('Authorization', '')
```
Comment (tomkralidis) [inline docker/auth/app.py:80]: auth_header = request.headers.get('Authorization')

File: docker/auth/app.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    if 'Bearer' in auth_header:
```
Comment (tomkralidis) [inline docker/auth/app.py:81]: `if auth_header is not None and if 'Bearer' in auth_header:`

File: docker/auth/app.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        api_key = auth_header.replace('Bearer', '').strip()
+
+    # check if resource passed exists in auth list
+    # if no, it's open, return
+    if resource == '' or is_resource_open(resource):
```
Comment (tomkralidis) [inline docker/auth/app.py:86]: `if resource is None or is_resource_open(resource):`

File: docker/auth/app.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    request_ = request.from_values(request_uri)
+
+    LOGGER.debug('Extracting topic from request URI')
+    resource = extract_topic(request_uri)
+
+    LOGGER.debug('Extracting API token')
+    api_key = request_.args.get('token')
+    auth_header = request.headers.get('Authorization', '')
+    if 'Bearer' in auth_header:
+        api_key = auth_header.replace('Bearer', '').strip()
```
Comment (tomkralidis) [inline docker/auth/app.py:82]: `api_key = auth_header.split()[-1].strip()`

File: docker/auth/app.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    :param uri: `str` of requested URI
+
+    :returns: `str` of API collection metadata
+    """
+
+    # TODO: Move away from regex matching for topic hierarchies.
+    pattern = r'(data[\/.][a-z]{4,11}[\/.][a-z-]+[\/.][a-z]{2}[\/.][A-Z]*[\/.][a-zA-Z]+)'  # noqa
+    prog = re.compile(pattern)
+    match = prog.search(uri)
+    return match[0] if match else ''
```
Comment (tomkralidis) [inline docker/auth/app.py:67]: `     return match[0] if match else None`

File: docker/default.env [hunk 6]
```diff
@@ -24,3 +24,6 @@ WIS2BOX_URL=http://localhost:8999
 # mappings of topic hierarchy to wis2box data plugins
 # optionally override default mappings from wis2box data plugins
 # WIS2BOX_DATADIR_DATA_MAPPINGS=${PWD}/wis2box-data-mappings.yml
+
+# access control
+WIS2BOX_AUTH_STORE=/data/wis2box/auth.db
```
Comment (tomkralidis) [inline docker/default.env:29]: Add this env var to the docs (https://wis2box.readthedocs.io/en/latest/configuration.html#other)

File: docs/source/auth.rst [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+A topic becomes open and no longer requires authentication when all tokens have been deleted. This can be done by
+deleting individual tokens, or all tokens for a given topic hierarchy.
+
+.. code-block:: bash
+
+    wis2box auth remove-tokens --topic-hierarchy data.core.observations-surface-land.mw.FWCL.landFixed
+    wis2box auth show
+
+
```
Comment (tomkralidis) [inline docs/source/auth.rst:53]: Add a section demonstrating command line curl requests (200 and 401).

File: wis2box/auth/base.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+        :param key: `str` key
+        :param th: `str` topic hierarchy
+
+        :returns: `bool` of result
+        """
+        salt, pw_hash = hash_new_password(key)
+        try:
+            with SQLite3Backend(self.db) as conn:
+                s = 'INSERT INTO auth (salt, key, topic) VALUES (?, ?, ?)'
```
Comment (tomkralidis) [inline wis2box/auth/base.py:179]: ensure parameter binding is consistent across CRUD functionality.
